File: 18. Muestro Thompson.R
## Muestreo Thomposon ##


# libraries ---------------------------------------------------------------

library(tidyverse)


# read data ---------------------------------------------------------------

datos <- readr::read_csv("data/Ads_CTR_Optimisation.csv")


# Algoritmo de muestreo de Thompson ---------------------------------------

d <- 10
N <- 10000

number_rewards_1 <- integer(d)
number_rewards_0 <- integer(d)

ads_selected <- integer(0)

total_rewards <- 0

for(n in 1:N){
  
  max_random <- 0
  ad <- 0
  
  for(i in 1:d){

    random_beta <- rbeta(n = 1, 
                         shape1 = number_rewards_1[i] + 1, 
                         shape2 = number_rewards_0[i] + 1)
    
    if(random_beta > max_random){
      
      max_random <- random_beta
      ad <- i
      
    }
    
  }
  
  ads_selected = append(ads_selected, ad)
  rewards <- datos[n, ad]
  
  
  if(rewards == 1){
    
    number_rewards_1[ad] <- number_rewards_1[ad] + 1
    
  }else{
    
    number_rewards_0[ad] <- number_rewards_0[ad] + 1
    
  }
  
  total_rewards = total_rewards + rewards
  
}

# visualización de resultados ---------------------------------------------

data.frame(ads_selected) %>% 
  ggplot2::ggplot(aes(ads_selected)) +
  geom_bar(fill = "darkred") +
  labs(title = "Frecuencia de anuncios",
       x = "Anuncio",
       y = "Frecuencia") +
  theme_bw()

File: 18. Muestro Thompson.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jan 11 20:30:54 2022

@author: Admin
"""

## Muestreo de Thompson ##

# =============================================================================
# libraries
# =============================================================================

import pandas as pd
import matplotlib.pyplot as plt
import random


# =============================================================================
# read data
# =============================================================================

data = pd.read_csv("data/Ads_CTR_Optimisation.csv")


# =============================================================================
# Algoritmo Thompson
# =============================================================================


N = 10000
d = 10

number_rewards_1 = [0] * d
number_rewards_0 = [0] * d

ads_selected = []
total_reward = 0


for n in range(0, N):
    
    max_random = 0
    
    ad = 0
    
    for i in range(0, d):
        
        random_beta = random.betavariate(number_rewards_1[i] + 1, 
                                         number_rewards_0[i] + 1)
        
        if random_beta > max_random:
            
            max_random = random_beta
            
            ad = i
            
    ads_selected.append(ad)
    reward = data.values[n, ad]
        
    if reward == 1:
            
        number_rewards_1[ad] = number_rewards_1[ad] + 1
            
    else:
            
        number_rewards_0[ad] = number_rewards_0[ad] + 1
        
    total_reward = total_reward + reward


# =============================================================================
# Visualización de resultados
# =============================================================================

plt.hist(ads_selected)
plt.title("Histograma de anuncios")
plt.xlabel("ID del anuncio")
plt.ylabel("Frecuencia de visualización")
plt.show()